src/software/embedded/setup_robot_software_deps.sh

```diff
@@ -14,3 +14,15 @@ sudo apt-get install "${host_software_packages[@]}" -y
 
 # Install platformio udev rules
 curl -fsSL https://raw.githubusercontent.com/platformio/platformio-core/develop/platformio/assets/system/99-platformio-udev.rules | sudo tee /etc/udev/rules.d/99-platformio-udev.rules
+
+# Install Python 3.12
+wget -N https://www.python.org/ftp/python/3.12.0/Python-3.12.0.tar.xz -O /tmp/tbots_download_cache/python-3.12.0.tar.xz
+tar -xf /tmp/tbots_download_cache/python-3.12.0.tar.xz -C /tmp/tbots_download_cache/
+cd /tmp/tbots_download_cache/Python-3.12.0
+./configure --enable-optimizations
+make -j$(nproc)
+sudo make altinstall
```

Comment:
Is compiling from source necessary?

Reply:
The Pis run Debian so `add-apt-repository` doesn't exist and we can't add `deadsnakes`. Maybe there's a .deb so that we can use but I couldn't find one

Reply:
Honestly, there are way too many compiling from source. 

Moving on next year, it may be just better to just have one person compile from source for everything, package it and put it on a cloud server?

Reply:
Compilation from source is honestly the easiest way. It's the best way because:
1. We support multiple Ubuntu versions with multiple GLIBC versions
2. We are supporting both ARM and X86 systems
3. According to CI logs, the setup scripts only take ~5 minutes to run on a 2-core machine, which isn't really that bad at all

Reply:
The reasons I suggested this is because there was a awkward time I have to switch between two branches requires me to run `seutp_software.sh` every time. The ~5 minutes of time added every time a branch is switched is painfully slow. If we switch 4 times that would be around 20 minutes of time. 

(1) multiple version of GLIBC is likely isn't an issue because it is backward compatible. GNU spend a lot of time making sure the ABI doesn't change that often so that people don't have to recompile. [see here.](https://abi-laboratory.pro/?view=timeline&l=glibc ) We can just compile using a older version of GLIBC. It is really unlikely that the few binaries dependencies we need just so happens to have changes in the ABI. 

(2) you could just make two deb files instead to support both ARM and X86. 

(3) 5 minutes is ok as long as it is ran once. Sometimes when rolling back to older commits, this becomes a problem. 

Finally, if you were to go with the deb route, or tarball route.  It would be so much easier to get the `cc_library` to link to a arm version of `libpython.so` or`libpython.a` . I assume the reason why `cc_library` wasn't linked to `libpython.so` is because you need to get a cross compiler to be set up  on `setup_software.sh` which is going to be a pain.

Reply:
1. Yeah we are in the transition period between different support versions so I feel that setup_software pain. Hopefully, once we're out of this transition period, we should be okay for the next few years with Ubuntu 24.

2. Yup fair enough, we could self-host our dependencies on the External Deps github compiled with the lowest GLIBC version we want to support. If we do this though, we should try to have good documentation on _how_ to do it for future maintainers.

Reply:
I'm assuming by the "deb/tarball" and libpython comment, you mean that we can package a cross--compiled ARM libpython?